Use `pkl-gha` for GitHub Action Workflows

StefMa opened the pull request:
Hey there 👋 

as a pkl-community you might be interested in using `pkl` "everywhere" 🙃 
So I played over the weekened a bit and introduce `pkl-gha` (pkl for github actions) to this project.

I don't mind if gets rejected. 
But I just thought as pkl "fanatics" I thought it make sense to use it here 😁 

Maybe todos:
* Add a github action that checks if the pkl files got formatted (in case someone didn't setup the githook correctly)
* Fix dependabot things. See also https://github.com/StefMa/pkl-gha/issues/4#issuecomment-2306325459

jasongwartz commented:
I love the idea!

I don't mind commit hooks as a developer convenience, but they're insufficient to guarantee consistency. Could you please add a new CI check that verifies that the generated YAML config is up-to-date with the Pkl source (like the TS `dist` one)?

jasongwartz commented on .githooks/pre-commit:
```diff
@@ -0,0 +1,10 @@
+
+for file in $files; do
+	if [ "${file##*.}" = "pkl" ]; then
+		pkl eval "$file" -o "${file%.pkl}.yml"
```
What happens if someone doesn't have a global install of `pkl` on their machine?

StefMa replied:
It would fail and the user can't commit 🫠

jasongwartz replied:
Haha sure, I know how commit hooks work - I mean, `-bash: pkl: command not found` is a very weird error message to receive when you run `git commit`, not the best contributor experience.

jasongwartz commented on .githooks/pre-commit:
I wonder whether rather than doing the transform in the pre-commit hook, this hook should verify no diff in the generated file and error if there is one; ie. force the developer to conciously re-gen and commit manually.

That would be consistent with the TS -> JS code and the `npm run package` shortcut in this repo.

StefMa replied:
I wonder if we even want to have a git hook for this if we have an Action that tests that? 🤔 

I also don't like the idea because the user has to "find out" how to update the yml files.
With the approach right now we give the user something "at hand" and show them how to update the Action files.

StefMa replied:
Got it.
Just to understand it correct, you suggest to add a script to the `package.json` that executes the command? 🤔 
And users have to execute `npm run package` for the `dist` as well as for the pkl actions (like `npm run pkl-actions` (or so) to update the actions?!

Is that what you want?

jasongwartz commented on .github/workflows/check-dist.pkl:
```diff
@@ -0,0 +1,74 @@
+    name = "Check Dist"
+    `runs-on` = "ubuntu-latest"
+    steps {
+      new {
+        name = "Checkout"
+        uses = "actions/checkout@v4"
+      }
+      new {
+        name = "Setup Node.js"
+        uses = "actions/setup-node@v4"
+        with {
+          ["node-version-file"] = ".nvmrc"
+          ["cache"] = "npm"
+        }
+      }
+      new {
+        name = "Install Dependencies"
+        run = "npm ci"
+      }
```
A fun future improvement would be showing that this can be refactored into a utility module so that we have the same configuration for checkout and setting up node across all workflows.

StefMa replied:
Done here 👉  https://github.com/pkl-community/setup-pkl/pull/47/commits/54eb10e28c653def5c95b8fd4dba0e869501b472

jasongwartz commented:
We might also want to remove the dependabot config for GitHub Actions, since it would propose updates to the generated files, not the pkl sources: https://github.com/pkl-community/setup-pkl/blob/main/.github/dependabot.yml#L3

StefMa commented:
> We might also want to remove the dependabot config for GitHub Actions, since it would propose updates to the generated files, not the pkl sources: https://github.com/pkl-community/setup-pkl/blob/main/.github/dependabot.yml#L3

Actually I wouldn't do that.
With that we would miss all the automatic updates by dependabot.

Since dependabot runs on GitHub actions we can add chain a new workflow with that.
As soon as dependebot updates an action, we can automaticallty run another action that updates the pkl file.
That should work ™️ but I haven't tried it (yet).

For now I would manually update the pkl files.

jasongwartz commented:
> With that we would miss all the automatic updates by dependabot.

I'm not talking about disabling Dependabot altogether, just for the Actions part. I don't think it makes sense for it to open PRs to generated files.

jasongwartz commented:
Related, it would be helpful to have pkl render out the files as eg `ci.gen.yaml` or `ci.generated.yaml` to make it clear they are not to be edited, and to add a preamble that says "THIS IS GENERATED, DO NOT EDIT".

jasongwartz commented on .github/workflows/modules/Steps.pkl:
```diff
@@ -0,0 +1,40 @@
+  with = customWith
+}
+
+function installPkl(): GitHubAction.Step = new {
+  name = "Install pkl"
+  uses = "pkl-community/setup-pkl@v0"
+  with {
+    ["pkl-version"] = "0.26.3"
+  }
+}
```
There's no need to use a function for something like this in Pkl, you can just do:
```suggestion
installPkl: GitHubAction.Step = new {
  name = "Install pkl"
  uses = "pkl-community/setup-pkl@v0"
  with {
    ["pkl-version"] = "0.26.3"
  }
}
```

StefMa replied:
Done here https://github.com/pkl-community/setup-pkl/pull/47/commits/e3b40c36d8f4f654ce498dd774c616c01fe4dc29

jasongwartz commented on .github/workflows/modules/Steps.pkl:
```diff
@@ -0,0 +1,40 @@
+import "package://pkg.pkl-lang.org/github.com/stefma/pkl-gha/com.github.action@0.0.3#/GitHubAction.pkl"
+
+function checkout(customWith: Mapping<String, String|Number|Boolean>?): GitHubAction.Step = new {
+  name = "Checkout"
+  uses = "actions/checkout@v4"
+  with = customWith
+}
```
Similarly here, a function isn't necessary for this in Pkl (actually, functions are *usually* not necessary in Pkl!).

You can implement it like:
```suggestion
checkout: GitHubAction.Step = new {
  name = "Checkout"
  uses = "actions/checkout@v4"
}
```

And then a consumer can use it and amend the `with` value like:
```
steps {
  (checkout) {
    with = "my custom with"
  }
}
```

StefMa replied:
Done here https://github.com/pkl-community/setup-pkl/pull/47/commits/e3b40c36d8f4f654ce498dd774c616c01fe4dc29

StefMa commented:
Hey @jasongwartz 

I wanted to push this to the finish line 🙃 

This question is still open:
https://github.com/pkl-community/setup-pkl/pull/47#discussion_r1732848338
Would be grat if you can answer this.
Be note that I'm not that deep into the javascript/npm word.
I might need some guidance how to add this to the package.json or some clear instructions what is expected from your side.
Thanks!

Other than that, I also rereanged some things:
* I moved the pkl-workflows into `.github/pkl-workflows`. 
* I moved the git-hook to a normal bash screen inside `.github/pkl-workflows`
  * I also executed this as part of the github action check (`check-action.pkl`) 

Thats it I guess 🙃 

Can't wait to get this thing merged 😇

jasongwartz commented on .github/pkl-workflows/updateWorkflows.sh:
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+
+for file in *.pkl; do
+	pkl eval "$file" -o "../workflows/$(basename "${file/.pkl/.generated.yml}")"
+done
```
It's not necessary to use a for-loop here - Pkl can evaluate multiple input files at once and template the output filename.

This can be simplified to:
```suggestion
pkl eval *.pkl -o "../workflows/%{moduleName}.generated.yml")"
```

In which case it's simple enough it probably doesn't need to be in a script! This could go nicely in a package.json "script" shortcut. And then yes, we could make "npm run package" also run this command, so generating dist and these files are one step.

StefMa replied:
Wow! This is super nice. Thanks for leeting me know ❤️

StefMa replied:
Fixed here https://github.com/pkl-community/setup-pkl/pull/47/commits/ed98e005893f58126d96eb0f3e7b226fe6a79c27

jasongwartz commented:
Could you please incorporate the changes from https://github.com/pkl-community/setup-pkl/pull/41 ?

jasongwartz commented on .github/pkl-workflows/ci.pkl:
```diff
@@ -0,0 +1,66 @@
+  ["test-action"] {
+    strategy {
+      matrix {
+        ["os"] = new {
```
If this has been typed correctly in your base module, you shouldn't need `= new {}` to create a new list, the override syntax `["os"] { ... }` should work. The only reason to do that typically with a Listing would be if there are default values in the parent object and you want to "start fresh" with a new Listing. Let me go check the typing in your parent module.

jasongwartz replied:
Looking at https://github.com/StefMa/pkl-gha/blob/main/GitHubAction.pkl#L592, I think there's extra complexity because `matrix` is typed as `Listing<String|*Dynamic>`. Can it really be a Dynamic object?

jasongwartz replied:
Gotcha - to me that looks more like it could be `Mapping<String, String>` - remember that `Dynamic` can contain properties, entries, AND elements, eg:
```
new {
  "an entry"
  prop1 = "value"
  ["entry one"] = "value2"
}
```

Note that this is actually not representable in JSON or YAML at all (below is from the REPL):

```
pkl> x = new {
>   "an entry"
>   prop1 = "value"
>   ["entry one"] = "value2"
> }
pkl> (new JsonRenderer {}).renderDocument(x)
–– Pkl Error ––
Cannot render object with both elements and properties/entries as JSON.
Object: new Dynamic { prop1 = "value"; ["entry one"] = "value2"; "an entry" }
```

But that's an aside, not important to this PR. Just info for you and your library

StefMa replied:
Thanks for the hint!
Do you now how I could model this better/differently? 🤔

jasongwartz commented on .github/pkl-workflows/modules/Steps.pkl:
```diff
@@ -0,0 +1,39 @@
+  uses = "actions/checkout@v4"
+}
+
+installPkl: GitHubAction.Step = new {
```
This is only used once - I think it would then be better to just define it in-line in `check-actions.pkl`. Putting it here confused me thinking we uesd it multiple places in the repo, but we don't - and it's more confusing because in `ci.pkl`, we check that the "local" version of `setup-pkl` is working (ie. the same action but as defined in the source code at current checkout).

StefMa replied:
Fixed here https://github.com/pkl-community/setup-pkl/pull/47/commits/26f6b76daa8325c3cd2a8fd8378b08f9f4d29ca3

jasongwartz commented:
Thanks @StefMa ! Left a few other pointers.

Re:
> I might need some guidance how to add this to the package.json or some clear instructions what is expected from your side.

In `package.json`, there is a field called `"scripts"` - each of these are little one-liners that can be run with `npm run` - that's where eg. `npm run package` is defined.

So could you please add a new script there called something like "gen:actions", which runs the one-liner from my comment on the current bash script? (The script won't be necessary after that) And then you can modify the "npm run package" script to add `&& npm run gen:actions` (so that running `npm run package` does both 😄 )

jasongwartz commented:
Feel free to send me a message on Discord if you want a hand working through any of that! And sorry for the delay in getting back to you on this

StefMa commented:
@jasongwartz its ready for review again 🙃

jasongwartz reviewed (approved):
lgtm, thanks for the contribution!
